src/backend/access/heap/heapam.c

```diff
@@ -5987,6 +5999,9 @@ heap_abort_speculative(Relation relation, ItemPointer tid)
 		elog(ERROR, "attempted to kill a non-speculative tuple");
 	Assert(!HeapTupleHeaderIsHeapOnly(tp.t_data));
 
+	if (HeapTupleHeaderIsSpeculative(tp.t_data))
+		ReleaseBuffer(buffer);  /* NEON: release buffer pinned by heap_insert */
```

Comment:
Why would this be false? Because of the toast case. That's quite subtle, let's add a comment here for that:
```suggestion
    /*
     * NEON: release buffer pinned by heap_insert
     *
     * This function is also used on the toast tuples of an aborted speculative
     * insertion. For those, there is no token on the tuple, and we didn' t keep
     * the pin. 
      */
	if (HeapTupleHeaderIsSpeculative(tp.t_data))
		ReleaseBuffer(buffer);  
```